### PR TITLE
Jpa cache properties recipe

### PR DESCRIPTION
**Pull request by cjobinabo**
## What's changed?

Web capture_28-11-2023_12555_

## Anything in particular you'd like reviewers to focus on?
Ideally we would insert `<shared-cache-mode>` before the `<validation-mode>` and `<properties>` nodes. However, with `addOrUpdateChild` it gets appended to the end.

Also the test is actually correct. There is just a spacing issue that appears to be causing the failure.

## Anyone you would like to review specifically?
@timtebeek @joanvr

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

**Comment by joanvr** [comment]
Hi @cjobinabo, thanks for your contribution! I've done some cleanup fixing some IntelliJ hints and improvements in the code, and reformatted the test so it's more clear and consistent and passes the tests now (no tabs/spaces issues).
Instead of one big test with all the cases in a single xml, maybe it could be more clear if we split it into multiple tests, each one with one scenario of the combination of possibilities?
I think this way it would be easier to debug and fix the remaining issues on formatting.

**Comment by cjobinabo** [comment]
Thanks for taking a look, Joan. I can break up the test.

File: src/main/java/org/openrewrite/java/migrate/JpaCacheProperties.java [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+        List<Content> contents = new ArrayList<>();
+        for (Content content : parent.getContent()) {
+            if (content == orginalProp) {
+                contents.add(updatedProp);
+            } else {
+                contents.add(content);
+            }
+        }
+        return parent.withContent(contents);
+    }
```
Comment (joanvr) [inline src/main/java/org/openrewrite/java/migrate/JpaCacheProperties.java]: why those methods are outside of the visitor?

File: src/main/java/org/openrewrite/java/migrate/JpaCacheProperties.java [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+            public Xml.Tag getARefNode(Xml.Tag puNode) {
+                for (Xml.Tag node : puNode.getChildren()) {
+                    if ("validation-mode".equals(node.getName())) {
+                        return node;
+                    } else if ("properties".equals(node.getName())) {
+                        return node;
+                    }
+                }
+                return null;
+            }
```
[the remark this answers is not in the document]
Comment (timtebeek) [reply src/main/java/org/openrewrite/java/migrate/JpaCacheProperties.java]: as indicated we could in the future add something [like we've done for Maven pom elements](https://github.com/openrewrite/rewrite/blob/d4042a039fa687bd8b917a9a30bfddb69d0a1053/rewrite-maven/src/main/java/org/openrewrite/maven/MavenTagInsertionComparator.java) and defined a particular insertion order. I don't think we need to hold up this PR over that though, so let's see this through as-is for now.

**Comment by timtebeek** [comment]
Thanks a lot @cjobinabo ! Enjoy your time off and look forward to seeing you again in the new year!

**Comment by jepsar** [comment]
We are using Payara 6.2025.2 (bundled with EclipseLink). After migration the application does not start with this exception:
```
java.io.IOException: org.xml.sax.SAXParseException; lineNumber: 201; columnNumber: 24; Deployment descriptor file META-INF/persistence.xml in archive [classes].  cvc-complex-type.2.4.d: Invalid content was found starting with element 'shared-cache-mode'. No child element is expected at this point.
```
Should I open a new issue for this?

**Comment by timtebeek** [comment]
Probably best indeed; thanks for reporting your findings!

**Pull request merged**